picard/coverart/providers/__init__.py

```diff
@@ -17,23 +17,74 @@
 # along with this program; if not, write to the Free Software
 # Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
 
+from picard import log, config
 from picard.plugin import ExtensionPoint
+from picard.ui.options import OptionsPage, register_options_page
 
 
 _cover_art_providers = ExtensionPoint()
 
 
+class ProviderOptions(OptionsPage):
+
+    """
+        Template class for provider's options
+    """
+
+    PARENT = "cover"
+    SORT_ORDER = 0
+    ACTIVE = True
+
+    options = []
+
+    _options_ui = None
+
+    def __init__(self, parent=None):
+        super(ProviderOptions, self).__init__(parent)
+        if callable(self._options_ui):
+            self.ui = self._options_ui()
+            self.ui.setupUi(self)
+
+    def load(self):
+        pass
+
+    def save(self):
+        pass
+
+
 def register_cover_art_provider(provider):
     _cover_art_providers.register(provider.__module__, provider)
+    if hasattr(provider, 'OPTIONS') and provider.OPTIONS:
+        provider.OPTIONS.NAME = provider.NAME
+        provider.OPTIONS.TITLE = provider.TITLE or provider.NAME
+        register_options_page(provider.OPTIONS)
 
 
 def cover_art_providers():
+    order = [p[0] for p in config.setting['ca_providers']]
+
+    def _key_provider(p):
+        try:
+            return order.index(p.NAME)
+        except ValueError:
+            return 666 # move to the end
     providers = []
-    for p in _cover_art_providers:
+    for p in sorted(_cover_art_providers, key=_key_provider):
         providers.append(p)
+    log.debug("CA Providers order: %s",
+              ' > '.join([p.NAME for p in providers]))
     return providers
 
 
+def is_provider_enabled(provider_name):
```

Comment:
Neither `CoverArtProvider.enabled` and `is_provider_enabled` document how they differ, although both from their contexts they both seem to return whether a cover art provider is enabled.

This is only used when a provider object is already in scope, so this could be a method on `CoverArtProvider`.